adding python version of STARK by Hand explainer

pdg744 opened the pull request:
This PR adds a python implementation of the [STARK by Hand](https://dev.risczero.com/proof-system/stark-by-hand) explainer. I wrote this with help from Tim and Andrew many months ago and kinda forgot about it. 

It would be nice if the code were a bit cleaner, but I think it makes sense to make this public and create one or more issues inviting improvements.

tzerrell commented on static/didactic_stark.py:
Can we make it really apparent what `f_super_3` and `g_super_(3)` are, and print out some sort of "success -- not rejected!" message when `f3 @ g(3) == mult`?

When I currently run the example, the end just says
```
Reject if f_super_3 evaluated @ g_super_(3) is not equal to 37
```
and I have to scroll up a bit to see
```
Choose g_super_(3) = 5^ 25  =  13 .
```
and
```
FRI Commitment (aka f_super_( 3 ): [37 37 37 37]
```
Having those right at the end, and making clear how that `13` index is in bounds for `f_super_(3)` (presumably its part of the evaluation domain?), would give a more of an immediate sense of success.

tzerrell commented on static/didactic_stark.py:
Cleanup printout
```suggestion
        print ("\n--- begin round", total_rounds - rounds_remaining + 1, "---")
```

tzerrell commented on static/didactic_stark.py:
Cleanup formatting
```suggestion
    print ("Choose g_super_(3) = 5^{} = {}.\n".format(g_super_3_index, g_super_3))
```

tzerrell commented on static/didactic_stark.py:
Cleanup formatting
```suggestion
    print ("Then, C_super_(2) = {{ g : g/5 is a square root of (g_super_3)/5 }} = {{ 5^{}, 5^{} }}".format(
        (g_super_3_index-1)//2 + 1,
        (g_super_3_index-1)//2 + (_field_size+1)//2
    ))
```

tzerrell reviewed (commented):
Getting this published instead of permanently tabled sounds good! I don't think we should have this as a dangling page -- at a minimum, we should link to it from STARK-by-hand page.

Additionally, its written in a fairly outdated structure -- it still references the four phase structure (trace, constraints, DEEP, FRI) which I believe we have fully removed as a current pedagogical approach. If you're thinking of revising the STARK By Hand article soon, we can hold off on changing this (and do them in tandem), but otherwise, I suggest we minimally drop the "PHASE" guideposts and perhaps also add "Lesson" guideposts.

I've made a few more inline suggestions, but I don't want them to slow this down too much -- let's favor land-and-revise over bogging down in revisions.

pdg744 commented:
Good call on the final terminal print-out -- I added the FRI commit & FRI query stuff as I was working on the FRI mechanics study club and my talk for zkLisbon; definitely needed some polishing on how that info was displayed!

pdg744 commented:
Have now cleaned up the final terminal print-out; I'm going to adjust the `Phase 1-4` language to align with the STARK by Hand explainer and then will ask for another review

pdg744 commented:
ok ready for re-review! 
- removed the outdated phase descriptions and 
- added a link from the STARK by Hand explainer to this Python STARK. 

Perhaps a bit sketch that the path to the python file is a direct download link; mb there's a better way to be making this available...?

tzerrell reviewed (approved):
Great, these revisions make this a much friendlier experience!

I think having the python version directly link to a file is reasonable -- we could wrap it as a markdown document with a ```python codeblock, but then actually running the file gets more annoying to do, and I think this will be much more often run than read as source code. I think the ideal UI would be some sort of in-browser live python notebook, which definitely exists, but not sure it's worth the effort to set up, at least right now.
